store/pg_engine_test.go

```diff
@@ -47,3 +59,66 @@ func TestGetMigrationVersions(t *testing.T) {
 		require.Equal(t, test.want, migrateVersions)
 	}
 }
+
+var (
+	pgUser        = "test"
+	pgPort        = 6000
+	serverVersion = "server-version"
+	l             = zap.NewNop()
+)
+
+func TestMigrationCompatibility(t *testing.T) {
+	pgDir := t.TempDir()
+	pgInstance, err := postgres.Install(path.Join(pgDir, "resource"), path.Join(pgDir, "data"), pgUser)
+	require.NoError(t, err)
+	err = pgInstance.Start(pgPort, os.Stdout, os.Stderr)
+	require.NoError(t, err)
+	defer pgInstance.Stop(os.Stdout, os.Stderr)
+
+	ctx := context.Background()
+	connCfg := dbdriver.ConnectionConfig{
+		Username: pgUser,
+		Password: "",
+		Host:     common.GetPostgresSocketDir(),
+		Port:     fmt.Sprintf("%d", pgPort),
+	}
+	d, err := dbdriver.Open(
+		ctx,
+		dbdriver.Postgres,
+		dbdriver.DriverConfig{Logger: l},
+		connCfg,
+		dbdriver.ConnectionContext{},
+	)
+	require.NoError(t, err)
+
+	err = d.SetupMigrationIfNeeded(ctx)
+	require.NoError(t, err)
+
+	versions, err := getMigrationFileVersions()
+	require.NoError(t, err)
+
+	// For every version, we create a database with the schema of that version and apply migrations till the latest version in the migration directory.
+	// For example, we can 3 versions, 1.0.0, 1.1.0, 1.2.0.
+	// Create a database with 1.0.0 schema, apply 1.1.0 migration, and apply 1.2.0 migration.
+	// Create a database with 1.1.0 schema, and apply 1.2.0 migration.
+	// Create a database with 1.2.0 schema. But there is no migration since it's the latest.
+	for i := range versions {
+		initialVersion := versions[i]
+		initialDatabaseName := getDatabaseName(initialVersion)
+		// Passing curVers = nil will create the database.
+		err = migrate(ctx, d, nil, initialVersion, serverVersion, initialDatabaseName, l)
+		require.NoError(t, err)
+
+		currentVersion := initialVersion
+		for j := i + 1; j < len(versions); j++ {
+			version := versions[j]
+			err = migrate(ctx, d, &currentVersion, version, serverVersion, initialDatabaseName, l)
```

Comment:
We may think a way to return the latest schema version and verify it. Ok to do in a separate PR

Reply:
Yeah. That was my plan but there's some difficulty because version X applying DDL to X+1 may have "different" schema of X+1 because of ordering and auto-generated trigger names.

Reply:
We can return the schema **version number** instead of the schema as the first step.

Reply:
Done.